### PR TITLE
Migrate serialization test from boost to catch

### PR DESCRIPTION
**Pull request by jeffin143**
There you go, We completed all the test and one more boost dependency removed , Refer #2523 

done_test

### CODE REVIEW

**Comment by birm** [comment]
Looks like macos is having some issues with the binding tests. Does this look familiar to you?

**Comment by jeffin143** [comment]
> Looks like macos is having some issues with the binding tests. Does this look familiar to you?

There is a discussion ongoing on IRC, any help would be appreciated :)

**Comment by jeffin143** [comment]
Also it is not serialisation test :-> https://github.com/mlpack/mlpack/blob/bf837f950b728bf08b80ccc696c6bc0d296e54af/src/mlpack/bindings/python/CMakeLists.txt#L305

**Review by birm — APPROVED** [review]
I thought it may have been a knock on effect of macos-steps.yaml, but you didn't change that one.
Anyway, this is the end of an era, isn't it :fireworks:

**Review by zoq — COMMENTED** [review]
Wow, great job, before we merge this in we should make sure https://github.com/mlpack/mlpack/pull/2678 is merged first, so we don't delete the instructions before that.

**Comment by jeffin143** [comment]
yes @zoq , Also I added some docs on how to run test , and how to run single or all binding related test, let me know if that's not needed, I will remove it

cc @birm  : May be take a second pass over since I added some docs

File: src/mlpack/tests/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Mlpack Tests
```
Comment (zoq) [inline src/mlpack/tests/README.md:1]: Let's be consistent and use `mlpack` everywhere.

File: src/mlpack/tests/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+This directory contains code and data used to test all the algorthims and function implemented in mlpack.
```
Comment (zoq) [inline src/mlpack/tests/README.md:3]:
```suggestion
This directory contains code and data used to test all the algorithms and functions implemented in mlpack.
```

File: src/mlpack/tests/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+## Test Directories Structure
+
+- *_test.cpp - All the test realted to algorthims
+- main_tests/*_test.cpp - All binding realted test
+- data - Data needed to run the test
```
Comment (zoq) [inline src/mlpack/tests/README.md:9]:
```suggestion
- *_test.cpp - methods tests
- main_tests/*_test.cpp - binding tests
- data - data needed to run the tests
```

Simplification?

File: src/mlpack/tests/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+## Add tests 
+
+We have a rich test suite, consisting of almost 2000 test and still counting. So it is suggested to add tests when:
+
+- Adding new functionality.
+- Fixing regressions and bugs.
+
+## Building Test
+
+To build test you can simply run `make mlpack_test`.
```
Comment (zoq) [inline src/mlpack/tests/README.md:20]:
```suggestion
To build the test suite you can simply run `make mlpack_test`.
```

File: src/mlpack/tests/README.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+- data - Data needed to run the test
+
+## Add tests 
+
+We have a rich test suite, consisting of almost 2000 test and still counting. So it is suggested to add tests when:
+
+- Adding new functionality.
+- Fixing regressions and bugs.
+
+## Building Test
```
Comment (zoq) [inline src/mlpack/tests/README.md:18]:
```suggestion
## Building Tests
```

File: src/mlpack/tests/README.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+To build test you can simply run `make mlpack_test`.
+
+## To run Test
```
Comment (zoq) [inline src/mlpack/tests/README.md:22]:
```suggestion
## To run Tests
```

File: src/mlpack/tests/README.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+- Adding new functionality.
+- Fixing regressions and bugs.
+
+## Building Test
+
+To build test you can simply run `make mlpack_test`.
+
+## To run Test 
+
+We use `Catch2` to write our test and to run the test, you can simply run
```
Comment (zoq) [inline src/mlpack/tests/README.md:24]:
```suggestion
We use `Catch2` to write our tests, to run all tests, you can simply run:
```

File: src/mlpack/tests/README.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+` ./bin/mlpack_test`
```
Comment (zoq) [inline src/mlpack/tests/README.md:26]:
```suggestion
`./bin/mlpack_test`
```

Remove extra space.

File: src/mlpack/tests/README.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+Also to run all the test in a particular file you can issue a command as
```
Comment (zoq) [inline src/mlpack/tests/README.md:28]:
```suggestion
Also to run all tests in a particular file you can run:
```

File: src/mlpack/tests/README.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+`./bin/mlpack_test "[testname]"`
+
+for example to run all the test which is in `cv_test.cpp` file you can issue a command as
```
Comment (zoq) [inline src/mlpack/tests/README.md:32]:
```suggestion
where `testname` is the name of the test suite. For example to run all
collaborative filtering tests implemented in `cv_test.cpp` you can run:
```

Just a suggestion.

File: src/mlpack/tests/README.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+`./bin/mlpack_test "[CVTest]"`
+
+Now similary you can run all the binding related test as
```
Comment (zoq) [inline src/mlpack/tests/README.md:36]:
```suggestion
Now similarly you can run all the binding related tests using:
```

Minor spelling fix.

File: src/mlpack/tests/README.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+`./bin/mlpack_test "[CVTest]"`
+
+Now similary you can run all the binding related test as 
+
+`./bin/mlpack_test "[BindingTests]"`
+
+To run a single test you can explicitly give the name of the test, for example to run
+`BinaryClassificationMetricsTest` present in `cv_test.cpp` file you can run the following
+command:
```
Comment (zoq) [inline src/mlpack/tests/README.md:42]:
```suggestion

To run a single test, you can explicitly provide the name of the test, for example, to run
`BinaryClassificationMetricsTest` implemented in `cv_test.cpp` you can run the following:
```

File: src/mlpack/tests/README.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+`./bin/mlpack_test BinaryClassificationMetricsTest`
+
+Catch2 also allows very other command which you can view using:
```
Comment (zoq) [inline src/mlpack/tests/README.md:46]:
```suggestion
Catch2 provides many other features like filter, checkout the [Catch2 reference section](docs/Readme.md#top) - for more details.
```

With that, I would delete the section below.

**Review by zoq — COMMENTED** [review]
Just some suggestions for the documentation.

File: src/mlpack/tests/README.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+- Adding new functionality.
+- Fixing regressions and bugs.
+
+## Building Tests
+
+To build the test suite you can simply run `make mlpack_test`.
+
+## To run Tests
+
+We use `Catch2` to write our tests, to run all tests, you can simply run:
```
Comment (birm) [inline src/mlpack/tests/README.md:24]:
```suggestion
We use `Catch2` to write our tests. To run all tests, you can simply run:
```

File: src/mlpack/tests/README.md [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+`./bin/mlpack_test`
+
+Also to run all tests in a particular file you can run:
```
Comment (birm) [inline src/mlpack/tests/README.md:28]:
```suggestion
To run all tests in a particular file you can run:
```

File: src/mlpack/tests/README.md [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+## Test Directories Structure
+
+- *_test.cpp - methods tests
+- main_tests/*_test.cpp - binding tests
+- data - data needed to run the tests
+
+## Add tests 
+
+We have a rich test suite, consisting of almost 2000 test and still counting. So it is suggested to add tests when:
```
Comment (birm) [inline src/mlpack/tests/README.md:13]:
```suggestion
We have a rich test suite, consisting of almost 2000 tests (and still counting). It is suggested to add tests when:
```

**Review by birm — APPROVED** [review]
Minor doc readability suggestions, but the core still seems good :)

**Comment by Aakash-kaushik** [comment]
Goodbye to Boost unit_test_framework.

**Review by zoq — APPROVED** [review]
Wow, great job everyone.

tenor

**Pull request merged**